### PR TITLE
JBProjects tests

### PR DESCRIPTION
**Pull request by joax**
Test coverage up to 100% on JBProjects

image

### CODE REVIEW

File: test/jb_projects/challenge_handle.test.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+  beforeEach(async function () {
+    [deployer, projectOwner, ...addrs] = await ethers.getSigners();
+
+    jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+
+    jbProjectsFactory = await ethers.getContractFactory('JBProjects');
+    jbProjectsStore = await jbProjectsFactory.deploy(jbOperatorStore.address);
+  });
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:30]: We've adopted a new pattern for test setup. Refer to the JBDirectory tests for more details – trying to do away with cross-test stateful variables whenever possible. Replace in this file and all others.

File: test/helpers/utils.js [hunk 2]
```diff
@@ -1,6 +1,22 @@
 import { BigNumber } from '@ethersproject/bignumber';
 import { ethers, network } from 'hardhat';
 
+export async function getTimestampFn(block) {
```
Comment (odd-amphora) [inline test/helpers/utils.js:4]: Remove 'fn' from this function and the one below.

File: test/helpers/utils.js [hunk 3]
```diff
@@ -1,6 +1,22 @@
...
+
+export async function fastforwardFn(block, seconds) {
+  const now = await getTimestampFn();
+  const timeSinceTimemark = now.sub(await getTimestampFn(block));
+  const fastforwardAmount = seconds.toNumber() - timeSinceTimemark;
+
+  // Subtract away any time that has already passed between the start of the test,
+  // or from the last fastforward, from the provided value.
+  await ethers.provider.send('evm_increaseTime', [fastforwardAmount]);
+  // Mine a block.
```
Comment (odd-amphora) [inline test/helpers/utils.js:16]: comment is not needed

File: test/helpers/utils.js [hunk 4]
```diff
@@ -1,6 +1,22 @@
...
+  return ethers.BigNumber.from((await ethers.provider.getBlock(block || 'latest')).timestamp);
+};
+
+export async function fastforwardFn(block, seconds) {
+  const now = await getTimestampFn();
+  const timeSinceTimemark = now.sub(await getTimestampFn(block));
+  const fastforwardAmount = seconds.toNumber() - timeSinceTimemark;
+
+  // Subtract away any time that has already passed between the start of the test,
+  // or from the last fastforward, from the provided value.
```
Comment (odd-amphora) [inline test/helpers/utils.js:14]: it seems like such a comment should go above fastForwardAmount

File: test/jb_projects/challenge_handle.test.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+import { getTimestampFn } from '../helpers/utils'
+
+describe('JBProjects::challengeHandle(...)', function () {
+
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:6]: I've noticed this pattern where a space is left in between new blocks (in the tests as well) – for consistency let's try to align things the same.

File: test/jb_projects/challenge_handle.test.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  let jbProjectsStore;
+
+  let deployer;
+  let projectOwner;
+  let addrs;
+
+  let projectHandle = "PROJECT_1";
+  let projectHandleNotTaken = "PROJECT_2"
+  let projectHandleNotTaken2 = "PROJECT_3"
+  let metadataCid = "";
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:20]: These should be const and the variable names should be capitalized snake case.

PROJECT_HANDLE, PROJECT_HANDLE_NOT_TAKEN1, PROJECT_HANDLE_NOT_TAKEN2

File: test/jb_projects/challenge_handle.test.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+  let jbProjectsFactory;
+  let jbProjectsStore;
+
+  let deployer;
+  let projectOwner;
+  let addrs;
+
+  let projectHandle = "PROJECT_1";
+  let projectHandleNotTaken = "PROJECT_2"
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:18]: for consistency: PROJECT_HANDLE_NOT_TAKEN1

File: test/jb_projects/challenge_handle.test.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+
+    jbProjectsFactory = await ethers.getContractFactory('JBProjects');
+    jbProjectsStore = await jbProjectsFactory.deploy(jbOperatorStore.address);
+  });
+
+  // Working on these now
+
+  it('Challenging an inexistent projectId', async function () {
+
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:35]: example of space from comment above.

File: test/jb_projects/challenge_handle.test.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    [deployer, projectOwner, ...addrs] = await ethers.getSigners();
+
+    jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+
+    jbProjectsFactory = await ethers.getContractFactory('JBProjects');
+    jbProjectsStore = await jbProjectsFactory.deploy(jbOperatorStore.address);
+  });
+
+  // Working on these now
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:32]: remove

File: test/jb_projects/claim_handle.test.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+
+    jbProjectsFactory = await ethers.getContractFactory('JBProjects');
+    jbProjectsStore = await jbProjectsFactory.deploy(jbOperatorStore.address);
+  });
+
+  // Working on these now
+
+  it('Should reject with Unauthorized', async function () {
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:35]: Let's be descriptive in the test name – why should it be rejected?

**Review by odd-amphora — COMMENTED** [review]
Please run `yarn pretty`

File: test/jb_projects/create_for.test.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    await expect(tx)
+      .to.emit(jbProjectsStore, 'Create')
+      .withArgs(
+        1,
+        projectOwner.address,
+        ethers.utils.formatBytes32String(PROJECT_HANDLE),
+        METADATA_CID,
+        deployer.address,
+      );
+  });
```
[the remark this answers is not in the document]
Comment (odd-amphora) [reply test/jb_projects/create_for.test.js:86]: I notice that this is a pattern across the remaining tests – please double check that everything in the contract is validated. From what I can tell there are quite a few pieces missing.

[the remark this answers is not in the document]
Comment (joax) [reply test/jb_projects/create_for.test.js:86]: One thing, on the metadataCid.length == 0 there is already a test with that (empty metadataCid, previous to this case).

[the remark this answers is not in the document]
Comment (joax) [reply test/jb_projects/create_for.test.js]: @odd-amphora take a look at the changes on the last commit.

File: test/jb_projects/challenge_handle.test.js [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+    return {
+      projectOwner,
+      deployer,
+      addrs,
+      jbProjectsStore,
+    };
+  }
+
+  it("Doesn't challenge if inexistent projectId", async function () {
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:32]: nit) 'Can't challenge if project doesn't exist'

File: test/jb_projects/challenge_handle.test.js [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    const { projectOwner, deployer, jbProjectsStore } = await setup();
+
+    await jbProjectsStore
+      .connect(deployer)
+      .createFor(
+        /*owner=*/ projectOwner.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE),
+        /*METADATA_CID=*/ METADATA_CID,
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:40]: param arguments aren't really needed here (except maybe owner, arguable) - actually for most APIs. I think they should really only be used when it's ambiguous. I'd say remove them from all these tests, but if you decide to keep them, format them consistently – METADATA_CID -> metadataCid.

File: test/jb_projects/challenge_handle.test.js [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+      );
+
+    await jbProjectsStore
+      .connect(projectOwner)
+      .transferHandleOf(
+        /*projectId=*/ PROJECT_ID,
+        /*address=*/ deployer.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_NOT_TAKEN),
+      );
+
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:50]: Most of this setup seems irrelevant to this test? I would think we just need to challenge a unclaimed handle, no?

File: test/jb_projects/challenge_handle.test.js [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+    await expect(
+      jbProjectsStore
+        .connect(deployer)
+        .challengeHandle(/*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_NOT_TAKEN_2)),
+    ).to.be.revertedWith('0x0d: HANDLE_NOT_TAKEN');
+  });
+
+  it("Doesn't challenge if a handle that has been challenged before", async function () {
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:58]: nit) Can't challenge ...

File: test/jb_projects/challenge_handle.test.js [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    await jbProjectsStore
+      .connect(deployer)
+      .challengeHandle(/*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_NOT_TAKEN));
+
+    await expect(
+      jbProjectsStore
+        .connect(deployer)
+        .challengeHandle(/*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_NOT_TAKEN)),
+    ).to.be.revertedWith('0x0e: CHALLENGE_OPEN');
+  });
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:86]: the global PROJECT_HANDLE_NOT_TAKEN is admittedly awkward here.

also, per my comment above, is all of this setup really needed? why is the transfer needed at all? let's just create the project and challenge twice?

File: test/jb_projects/challenge_handle.test.js [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+        /*metadataCid=*/ '',
+      );
+
+    await jbProjectsStore
+      .connect(projectOwner)
+      .transferHandleOf(
+        /*projectId=*/ PROJECT_ID,
+        /*address=*/ deployer.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_NOT_TAKEN),
+      );
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:145]: Same re: test setup. Also, generally the happy path comes first in the tests we've written so far.

File: test/jb_projects/challenge_handle.test.js [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+  const PROJECT_HANDLE = 'PROJECT_1';
+  const PROJECT_HANDLE_NOT_TAKEN = 'PROJECT_2';
+  const PROJECT_HANDLE_NOT_TAKEN_2 = 'PROJECT_3';
+  const METADATA_CID = '';
+  const PROJECT_ID = 1;
+
+  async function setup() {
+    let [deployer, projectOwner, ...addrs] = await ethers.getSigners();
+
+    let mockJbOperatorStore = await deployMockContract(deployer, jbOperatoreStore.abi);
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:19]: I see no tests here that test permissions? This is crucial for every API, please add them.

File: test/jb_projects/challenge_handle.test.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+
+  it(`Should challenge handle successfully`, async function () {
+    const { projectOwner, deployer, addrs, jbProjectsStore } = await setup();
+
+    await jbProjectsStore
+      .connect(deployer)
+      .createFor(
+        /*owner=*/ projectOwner.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        /*metadataCid=*/ '',
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:38]: This should be `METADATA_CID`, and since the variable name is obvious it should not need a param comment.

File: test/jb_projects/challenge_handle.test.js [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+      .challengeExpiryOf(ethers.utils.formatBytes32String(PROJECT_HANDLE_1));
+    await expect(storedChallengeExpiryOf).equal(expectedChallengeExpiry);
+
+    await expect(tx)
+      .to.emit(jbProjectsStore, 'ChallengeHandle')
+      .withArgs(
+        ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        PROJECT_ID,
+        expectedChallengeExpiry,
+        addrs[0].address,
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:57]: please be consistent with param comments. first and fourth params would need comments here.

File: test/jb_projects/challenge_handle.test.js [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+
+  it(`Can't challenge if a handle that has been challenged before`, async function () {
+    const { projectOwner, deployer, jbProjectsStore } = await setup();
+
+    await jbProjectsStore
+      .connect(deployer)
+      .createFor(
+        /*owner=*/ projectOwner.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        /*metadataCid=*/ METADATA_CID,
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:79]: param comments for handle and metadataCid not needed (in other places in this file, too.

File: test/jb_projects/claim_handle.test.js [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+import { deployMockContract } from '@ethereum-waffle/mock-contract';
+
+import jbOperatoreStore from '../../artifacts/contracts/JBOperatorStore.sol/JBOperatorStore.json';
+
+describe('JBProjects::claimHandle(...)', function () {
+  const PROJECT_HANDLE_1 = 'PROJECT_1';
+  const PROJECT_HANDLE_2 = 'PROJECT_2';
+  const PROJECT_HANDLE_3 = 'PROJECT_3';
+  const METADATA_CID = '';
+  const PROJECT_ID = 1;
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:15]: nit) PROJECT_ID_1

File: test/jb_projects/claim_handle.test.js [hunk 23]
```diff
@@ -0,0 +1,297 @@
...
+  const PROJECT_ID_2 = 2;
+
+  let jbOperatorStore;
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:18]: why is this jbOperatorStore here?

File: test/jb_projects/challenge_handle.test.js (hunk 19 above)
Comment (odd-amphora) [reply test/jb_projects/challenge_handle.test.js:38]: similarly `handle` is not needed as a comment as well. It's clear from PROJECT_HANDLE_1 and that factor it's just formatting a string.

File: test/jb_projects/claim_handle.test.js [hunk 24]
```diff
@@ -0,0 +1,297 @@
...
+
+  it(`Should claim handle of a project`, async function () {
+    const { projectOwner, deployer, jbProjectsStore } = await setup();
+
+    await jbProjectsStore
+      .connect(deployer)
+      .createFor(
+        /*owner=*/ projectOwner.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        /*metadataCid=*/ METADATA_CID,
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:55]: same thing here and in this file

File: test/jb_projects/claim_handle.test.js [hunk 25]
```diff
@@ -0,0 +1,297 @@
...
+    await expect(storedId).equal(PROJECT_ID_2);
+    await expect(storedHandle).equal(ethers.utils.formatBytes32String(PROJECT_HANDLE_1));
+
+    await expect(tx)
+      .to.emit(jbProjectsStore, 'ClaimHandle')
+      .withArgs(
+        PROJECT_ID_2,
+        deployer.address,
+        ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        deployer.address,
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:96]: param comments for second and fourth params

File: test/jb_projects/create_for.test.js [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+  const METADATA_CID = '';
+  const PROJECT_ID = 1;
+  const PROJECT_ID_2 = 2;
+
+  let jbOperatorStore;
+
+  beforeEach(async function () {
+    let jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+  });
```
Comment (odd-amphora) [inline test/jb_projects/create_for.test.js:17]: This shouldn't be in the beforeEach - it should be in setup.

File: test/jb_projects/claim_handle.test.js [hunk 27]
```diff
@@ -0,0 +1,297 @@
...
+  const PROJECT_ID = 1;
+  const PROJECT_ID_2 = 2;
+
+  let jbOperatorStore;
+  let CLAIM_HANDLE_PERMISSION_INDEX;
+
+  beforeEach(async function () {
+    let jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
+
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:24]: operator store should be removed from beforeEach since it is stateful. the operations/permission index can just be wrapped in before(async... because it won't change.

File: test/jb_projects/create_for.test.js [hunk 28]
```diff
@@ -0,0 +1,140 @@
+import { expect } from 'chai';
+import { ethers } from 'hardhat';
+
+describe('JBProjects::createFor(...)', function () {
+  const PROJECT_HANDLE_1 = 'PROJECT_1';
+  const PROJECT_HANDLE_2 = 'PROJECT_2';
+  const PROJECT_HANDLE_EMPTY = '';
+  const METADATA_CID = '';
+  const PROJECT_ID = 1;
```
Comment (odd-amphora) [inline test/jb_projects/create_for.test.js:9]: nit) PROJECT_ID_1

File: test/jb_projects/create_for.test.js [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+
+  it(`Should create a project`, async function () {
+    const { projectOwner, deployer, jbProjectsStore } = await setup();
+
+    let tx = await jbProjectsStore
+      .connect(deployer)
+      .createFor(
+        /*owner=*/ projectOwner.address,
+        /*handle=*/ ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        /*metadataCid=*/ METADATA_CID,
```
Comment (odd-amphora) [inline test/jb_projects/create_for.test.js:41]: same thing.

File: test/jb_projects/create_for.test.js [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+    await expect(storedMetadataCid).to.equal(METADATA_CID);
+
+    await expect(tx)
+      .to.emit(jbProjectsStore, 'Create')
+      .withArgs(
+        PROJECT_ID,
+        projectOwner.address,
+        ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        METADATA_CID,
+        deployer.address,
```
Comment (odd-amphora) [inline test/jb_projects/create_for.test.js:61]: same thing - when relevant.

File: test/jb_projects/renew_handle_of.test.js [hunk 31]
```diff
@@ -0,0 +1,132 @@
...
+  const PROJECT_HANDLE = 'PROJECT_1';
+  const METADATA_CID = '';
+  const PROJECT_ID = 1;
+
+  let jbOperatorStore;
+  let RENEW_HANDLE_PERMISSION_INDEX;
+
+  beforeEach(async function () {
+    let jbOperatorStoreFactory = await ethers.getContractFactory('JBOperatorStore');
+    jbOperatorStore = await jbOperatorStoreFactory.deploy();
```
Comment (odd-amphora) [inline test/jb_projects/renew_handle_of.test.js:18]: same thing.

**Review by odd-amphora — COMMENTED** [review]
The same comments apply to all files.

File: test/jb_projects/transfer_handle_of.test.js [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+    await expect(
+      jbProjectsStore
+        .connect(deployer)
+        .transferHandleOf(
+          PROJECT_ID_1,
+          deployer.address,
+          ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        ),
+    ).to.be.reverted;
+  });
```
Comment (odd-amphora) [inline test/jb_projects/transfer_handle_of.test.js:194]: these two tests have the same name?

File: test/jb_projects/claim_handle.test.js [hunk 33]
```diff
@@ -0,0 +1,296 @@
...
+        .connect(deployer)
+        .claimHandle(
+          ethers.utils.formatBytes32String(PROJECT_HANDLE_2),
+          deployer.address,
+          PROJECT_ID_2,
+        ),
+    ).to.be.revertedWith('0x0c: UNAUTHORIZED');
+  });
+
+  if (
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:211]: it? and there's extra parenthesis it looks like

File: test/jb_projects/claim_handle.test.js [hunk 34]
```diff
@@ -0,0 +1,296 @@
...
+            .claimHandle(
+              ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+              projectOwner.address,
+              PROJECT_ID_2,
+            ),
+        ).to.be.reverted;
+      })
+  );
+
+  if (
```
Comment (odd-amphora) [inline test/jb_projects/claim_handle.test.js:254]: here as well

File: test/jb_projects/challenge_handle.test.js [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+
+    return {
+      projectOwner,
+      deployer,
+      addrs,
+      jbProjectsStore,
+    };
+  }
+
+  it(`Should challenge handle successfully`, async function () {
```
Comment (odd-amphora) [inline test/jb_projects/challenge_handle.test.js:30]: name) For all the happy paths let's add "and emit event(s)" to the name

**Review by odd-amphora — APPROVED** [review]

File: test/jb_projects/claim_handle.test.js [hunk 36]
```diff
@@ -0,0 +1,290 @@
...
+        .connect(handleReceiver)
+        .claimHandle(
+          ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+          projectOwner.address,
+          PROJECT_ID_2,
+        ),
+    ).to.be.reverted;
+  });
+
+  it(`Can't claim handle even if its recipient who does it but without permissions`, async function () {
```
Comment (mejango) [inline test/jb_projects/claim_handle.test.js:251]: not sure about this test.

The two `requirePermission`'s block for `claimHandle` make sure that the address claiming is both the address that the handle was transferred to, and the owner of the project that the claimed handle is being set for.

File: test/jb_projects/claim_handle.test.js [hunk 37]
```diff
@@ -0,0 +1,290 @@
...
+      )
+      .returns(false);
+
+    await expect(
+      jbProjectsStore
+        .connect(handleReceiver)
+        .claimHandle(
+          ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+          projectOwner.address,
+          PROJECT_ID_2,
```
Comment (mejango) [inline test/jb_projects/claim_handle.test.js:286]: this is failing because handleReceiver isn't the owner of PROJECT_ID_2

File: test/jb_projects/transfer_handle_of.test.js [hunk 38]
```diff
@@ -0,0 +1,173 @@
...
+        .connect(addrs[0])
+        .transferHandleOf(
+          PROJECT_ID_1,
+          deployer.address,
+          ethers.utils.formatBytes32String(PROJECT_HANDLE_1),
+        ),
+    ).to.be.reverted;
+  });
+
+  it(`Can't transfer handle if operator even with permissions`, async function () {
```
Comment (mejango) [inline test/jb_projects/transfer_handle_of.test.js:148]: this one also doesnt make sense to me

File: test/jb_projects/transfer_handle_of.test.js [hunk 39]
```diff
@@ -0,0 +1,173 @@
...
+      );
+
+    await mockJbOperatorStore.mock.hasPermission
+      .withArgs(deployer.address, projectOwner.address, PROJECT_ID_1, SET_HANDLE_PERMISSION_INDEX)
+      .returns(true);
+
+    await expect(
+      jbProjectsStore
+        .connect(deployer)
+        .transferHandleOf(
```
Comment (mejango) [inline test/jb_projects/transfer_handle_of.test.js:166]: this is failing because the `_newHandle` is PROJECT_HANDLE_1 which is the same as the current handle for the project with PROJECT_ID_1. The error here is 0x0b: HANDLE_TAKEN

**Review by mejango — COMMENTED** [review]
a few curiosities

File: test/jb_projects/claim_handle.test.js (hunk 36 above)
Comment (joax) [reply test/jb_projects/claim_handle.test.js:251]: Correct, but what if the handle is transferred and later the permission is removed? we need to test that when the permission is revoked, the recipient of the handle cannot claim it.

Comment (mejango) [reply test/jb_projects/claim_handle.test.js:251]: I still think we're slightly off the same page.

`requirePermission(_transferAddress, _projectId, JBOperations.CLAIM_HANDLE)` this line means "the msg.sender must either be the _transferAddress, or it must have been given the JBOperations.CLAIM_HANDLE permission from the _transferAddress".

we need to test that when the permission is given to an operator by the _transferRecipient, the operator can claim it the handle if it is also the owner of project having the handle claimed onto or an operator thereof. The _transferRecipient can also claim at any time if it is also the project owner of the handle being claimed onto or is an operator.

File: test/jb_projects/claim_handle.test.js (hunk 37 above)
Comment (joax) [reply test/jb_projects/claim_handle.test.js:294]: Correct, this is fixed.

File: test/jb_projects/transfer_handle_of.test.js (hunk 38 above)
Comment (joax) [reply test/jb_projects/transfer_handle_of.test.js]: You are right, the test need to be opposite (without permissions). Fixed.

File: test/jb_projects/transfer_handle_of.test.js (hunk 39 above)
Comment (joax) [reply test/jb_projects/transfer_handle_of.test.js:166]: You are right, it is fixed now. thx.

File: test/jb_projects/claim_handle.test.js (hunk 36 above)
Comment (joax) [reply test/jb_projects/claim_handle.test.js:251]: So what is missing is the operator's permissions test, right? the other two tests are already covered with previous tests.

Comment (joax) [reply test/jb_projects/claim_handle.test.js]: It happens I was initializing JBProjects without the mock contract for permissions mocking, that was driving me crazy. Now it should cover the use cases of:

1. Test handleReceiver claim --> success
2. Test operator claim (not owner of handle or receiver) --> fail
3. Test operator with permissions (not owner of handle or receiver) --> success

**Review by mejango — APPROVED** [review]
sweet
